tribits/ci_support/CDashAnalyzeReportRandomFailures.py

```diff
@@ -0,0 +1,339 @@
+        cdashSiteUrl, cdashProjectName, None, testHistoryQueryFilters)
+
+      testHistoryLOD = CDQAR.downloadTestsOffCDashQueryTestsAndFlatten(
+        testHistoryQueryUrl, testHistoryCacheFile, alwaysUseCacheFileIfExists=True)
+
+      print("\n  Size of test history: "+str(len(testHistoryLOD)))
+
+      if printUrlMode == 'all':
+        print("\n  CDash test history browser URL for "+nonpassingTest['testname']+" "+\
+          correctedBuildName+":\n\n"+"   "+testHistoryBrowserUrl)
+        print("\n  CDash test history query URL for "+nonpassingTest['testname']+" "+\
+          correctedBuildName+":\n\n"+"   "+testHistoryQueryUrl)
+
+      if len(testHistoryLOD) < 2:
+        print("\n  Size of test history too small for any comparisons, skipping ...\n")
+        continue
+
+      # B.3) Split full testing history to passed and nonpassed lists of dicts
+      passingTestHistoryLOD = [test for test in testHistoryLOD if test.get("status") == "Passed"]
+      nonpassingTestHistoryLOD = [test for test in testHistoryLOD if test.get("status") == "Failed"]
+      nonpassingSha1Pairs = set()
+
+      print("\n  Num of passing tests in test history: "+str(len(passingTestHistoryLOD)))
+      print("\n  Num of nonpassing tests in test history: "+str(len(nonpassingTestHistoryLOD)))
+
+      buildSummaryCacheDir = testQueriesCacheDir+"/build_summary_cache/"+testNameBuildName
+      createDirsFromPath(buildSummaryCacheDir)
+      # NOTE: There is an argument to be made that test histories should get their own directory
+      # instead of build summaries and that build summaries should live inside of there
+
+      # C.1) Get all nonpassing tests' sha1s into a set
+      for test in nonpassingTestHistoryLOD:
+
+        buildId = self.getBuildIdFromTest(test)
+
+        buildSummaryCacheFile = buildSummaryCacheDir+"/"+buildId
+        buildSummaryQueryUrl = CDQAR.getCDashBuildSummaryQueryUrl(cdashSiteUrl, buildId)
+        buildConfigOutput = self.downloadBuildSummaryOffCDash(
+          buildSummaryQueryUrl, buildSummaryCacheFile, verbose=printUrlMode =='all',
+          alwaysUseCacheFileIfExists=True)['configure']['output']
+
+        nonpassingSha1Pairs.add(
+          self.versionInfoStrategy.getTargetTopicSha1s(buildConfigOutput))
+
+      print("\n  Test history failing sha1s: "+str(nonpassingSha1Pairs))
+
+      # C.2) Check if passing tests' sha1s exist in nonpassing sha1s set
+      for test in passingTestHistoryLOD:
+
+        buildId = self.getBuildIdFromTest(test)
+
+        buildSummaryCacheFile = buildSummaryCacheDir+"/"+buildId
+        buildSummaryQueryUrl = CDQAR.getCDashBuildSummaryQueryUrl(cdashSiteUrl, buildId)
+        buildConfigOutput = self.downloadBuildSummaryOffCDash(
+          buildSummaryQueryUrl, buildSummaryCacheFile, verbose=printUrlMode =='all',
+          alwaysUseCacheFileIfExists=True)['configure']['output']
+
+        passingSha1Pair = \
+          self.versionInfoStrategy.getTargetTopicSha1s(buildConfigOutput)
```

Comment:
As I mentioned in my review of the file `test/ci_support/example_cdash_analyze_and_report_random_failures.py` above, a future refactoring should move all of this into a refactored strategy interface that takes `buildId` and returns the `versionInfo` object.  (And the `versionInfo` object just must be printable and it must be compatible.)   So the refactored code would be:

```
  passingVersionInfo = self.versionInfoStrategy.getTargetTopicSha1s(buildId)
```

Then other aways to get the "version" for a build for different projects could be devised, as I mention in my review of the file test/ci_support/example_cdash_analyze_and_report_random_failures.py` above.